### ISSUE TITLE
Warn on empty "body" for last pattern in case/cond/etc.

### ISSUE THREAD

**Issue by lexmag** [elixir-lang/elixir#3931]
```elixir
iex(1)> case 123 do
...(1)>   321 -> :error
...(1)>   123 ->
...(1)> end
nil
```

**Comment by eksperimental** [comment]
same with `cond`

```elixir
iex(4)> cond do
...(4)>   false -> :error
...(4)>   123 ->
...(4)> end
nil
end
```

**Comment by lexmag** [comment]
@eksperimental right, it's about `receive`, `try` as well.

**Comment by whatyouhide** [comment]
I'd love to take a stab at this. So far, I played around in `elixir_exp_clauses.erl` but I guess we have to do it at a lower level since `term ->` is still `{:->, _, [_, nil]}` at the AST level. This means that this:

```elixir
cond do
  true -> nil
end
```

would still raise the warning. Should we do this at the parsing level? I'm asking because it seems to me that there are no warnings issued in the parser, so it feels weird to do that.

**Comment by lexmag** [comment]
@whatyouhide the addition of `nil` happens in parser:
```elixir
iex(1)> str = ~C"""                                   
...(1)> case 123 do
...(1)>   321 -> :error
...(1)>   123 ->
...(1)> end
...(1)> """
'case 123 do\n  321 -> :error\n  123 ->\nend\n'
iex(2)> :elixir.string_to_quoted(str, 1, "nofile", [])
{:ok,
 {:case, [line: 1],
  [123,
   [do: [{:->, [line: 2], [[321], :error]}, {:->, [line: 3], ['{', nil]}]]]}}
```
I think we could mark this rewriting with some meta key and warn on it in the `elixir_exp.erl`, @josevalim what are your thoughts on it?

**Comment by josevalim** [comment]
We can warn in the parser. In case writing a warning is not straight-forward, let's just raise. The fix is backwards compatible anyway.

**Comment by whatyouhide** [comment]
Ok, after working on this a little more, I'm not sure how we should move given that stabs are parsed the same for both `fn`s and blocks.

`fn -> end` is legal, as is `fn ANYTHING_HERE -> end`. As far as I can understand from this issue's description, we only want to warn/raise in `cond`/`case`/... clauses, but not in `fn` clauses. If this is the case, I guess we have to distinguish in the parser between the `stab` used in `access_expr` (which allows empty right-hand sides) and the `stab` used in `do_block` expressions (which this way would never allow empty right-hand sides).

Playing with stabs, there are quite a lot of places where the behaviour is a little bit confusing, for example:

```elixir
(fn ->; ->; -> end).() #=> nil

cond do
  ->;
  ->
end #=> nil
```

**Comment by josevalim** [comment]
@whatyouhide the first case warns here. The second one does not compile. What am I missing? :)

**Comment by whatyouhide** [comment]
The first case (the `fn`) does not warn for me, neither in master nor on
v1.1.1. For the second case, my bad I wrote the wrong one :). I mean that
this doesn't issue any warnings:

```elixir
cond do
  1 ->;
  2 ->;
end
```

On Sunday, 1 November 2015, José Valim <notifications@github.com> wrote:

> @whatyouhide <https://github.com/whatyouhide> the first case warns here.
> The second one does not compile. What am I missing? :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/elixir-lang/elixir/issues/3931#issuecomment-152831206>
> .
>


-- 

Andrea Leopardi
an.leopardi@gmail.com

**Comment by josevalim** [comment]
@whatyouhide remember to always check for warnings by defining the code inside a function in a module. Eval'ed/shell code does not go through the regular compilation process!

Regarding the cond, both those cases are returning `nil`. The issue though is that we do support `fn -> end` and I am not sure if we should explicitly force a return value there. I guess we should anyway? /cc @lexmag

**Comment by whatyouhide** [comment]
Alright, rookie mistake. Yes the first one does warn. As for the second
one, yes I meant both cases return nil; my point was that this issue
extends not to just the last clause but to all clauses (if terminated so
that the parser separates correctly, like here using a semicolon).

My personal opinion is that we should ditch fn -> end as well, it's not
clear and very implicit. If we go with that route, then the change is
minimal (and I have it working locally) since we just warn/raise on
stab_expr in the parser.

On Sunday, 1 November 2015, José Valim <notifications@github.com> wrote:

> @whatyouhide <https://github.com/whatyouhide> remember to always check
> for warnings by defining the code inside a function in a module.
> Eval'ed/shell code does not go through the regular compilation process!
>
> Regarding the cond, both those cases are returning nil. The issue though
> is that we do support fn -> end and I am not sure if we should explicitly
> force a return value there. I guess we should anyway? /cc @lexmag
> <https://github.com/lexmag>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/elixir-lang/elixir/issues/3931#issuecomment-152832564>
> .
>


-- 

Andrea Leopardi
an.leopardi@gmail.com

**Comment by lexmag** [comment]
Yes, please, let's eliminate all the occurrences of that "magic".

**Comment by lexmag** [comment]
Closed in #3958.
